akka-stream/src/main/scala/akka/stream/scaladsl/FlowWithContext.scala

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2014-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.scaladsl
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.language.higherKinds
+import scala.annotation.unchecked.uncheckedVariance
+
+import akka.NotUsed
+import akka.dispatch.ExecutionContexts
+import akka.stream._
+import akka.stream.impl.LinearTraversalBuilder
+
+/**
+ * Describes the usage of a stream operator, passing through a context of type `Ctx`.
+ * The context is passed from input elements to output elements.
+ */
+sealed trait Op[+Ctx] {
+  def map[Ctx2](f: Ctx ⇒ Ctx2): Op[Ctx2]
+}
+
+/**
+ * Describes a one-to-one type operator, which passes the `context`
+ * from one input element to one output element.
+ */
+case class OneToOne[+Ctx](context: Ctx) extends Op[Ctx] {
+  def map[Ctx2](f: Ctx ⇒ Ctx2): Op[Ctx2] = OneToOne(f(context))
+}
+
+/**
+ * Describes a one-to-many type operator, which passes the `context`
+ * from one input element to many output elements.
+ */
+case class OneToMany[+Ctx](context: Ctx, last: Boolean = false) extends Op[Ctx] {
+  def map[Ctx2](f: Ctx ⇒ Ctx2): Op[Ctx2] = OneToMany(f(context))
+}
+
+/**
+ * Describes a many-to-one type operator, which passes the contexts
+ * from many input elements to one output element.
+ */
+case class ManyToOne[+Ctx](contexts: Seq[Ctx]) extends Op[Ctx] {
+  def map[Ctx2](f: Ctx ⇒ Ctx2): Op[Ctx2] = ManyToOne(contexts.map(f))
+}
+
+/**
+ * The contexts of many input elements is passed on to many output elements.
+ */
+case class ManyToMany[+Ctx](contexts: Seq[Ctx]) extends Op[Ctx] {
+  def map[Ctx2](f: Ctx ⇒ Ctx2): Op[Ctx2] = ManyToMany(contexts.map(f))
+}
+
+object Op {
+  /**
+   * Applies a many-to-many operation.
+   */
+  private[akka] def manyToMany[Ctx](contexts: Seq[Op[Ctx]]): ManyToMany[Ctx] = ManyToMany(contexts.flatMap {
+    case OneToOne(ctx)     ⇒ Seq(ctx)
+    case OneToMany(ctx, _) ⇒ Seq(ctx)
+    case ManyToOne(ctxs)   ⇒ ctxs
+    case ManyToMany(ctxs)  ⇒ ctxs
+  })
+
+  /**
+   * Applies a one-to-many operation.
+   */
+  private[akka] def oneToMany[Ctx](op: Op[Ctx]) = op match {
+    case OneToOne(ctx)        ⇒ OneToMany(ctx)
+    case otm: OneToMany[Ctx]  ⇒ otm
+    case ManyToOne(ctxs)      ⇒ ManyToMany(ctxs)
+    case mtm: ManyToMany[Ctx] ⇒ mtm
+  }
+}
+
+trait FlowWithContextOps[+Ctx, +Out, +Mat] {
+  type Repr[+C, +O] <: FlowWithContextOps[C, O, Mat] {
+    type Repr[+CC, +OO] = FlowWithContextOps.this.Repr[CC, OO]
+    type Prov[+CC, +OO] = FlowWithContextOps.this.Prov[CC, OO]
+  }
+
+  type Prov[+C, +O] <: FlowOpsMat[(O, Op[C]), Mat]
```

Comment:
Yep, seems we cannot get around this when we want to share code between SourceWithContext and FlowWithContext.